Mathlib/RingTheory/MvPolynomial/Symmetric.lean

```diff
@@ -309,4 +309,244 @@ theorem psum_isSymmetric (n : ℕ) : IsSymmetric (psum σ R n) := rename_psum _
 
 end PowerSum
 
+section Exponent
+
+open Classical
+open Composition
+open Finset
+
+variable (σ R) [CommSemiring R] [CommSemiring S] [Fintype σ] [Fintype τ]
+
+@[ext]
+structure Exponent (n : ℕ) :=
+  (exp : σ → ℕ)
+  (sum_eq : ∑ i, exp i = n)
+
+lemma Exponent.single_le_sum {n : ℕ} (s : Exponent σ n) (i : σ) : s.exp i ≤ n := by
+  have h : ∑ i, s.exp i = n := s.sum_eq
+  simp_rw [←h]
+  exact Finset.single_le_sum (fun i _ => Nat.zero_le (s.exp i)) (mem_univ i)
+  done
+
+lemma Exponent.zero_of_sum_eq_zero (s : Exponent σ 0) : ∀ i, s.exp i = 0 := by
+  intro i
+  have := Finset.single_le_sum (fun i _ => Nat.zero_le (s.exp i)) (mem_univ i)
+  rw [s.sum_eq] at this
+  exact Nat.eq_zero_of_le_zero this
+  done
+
+def Exponent.zero : Exponent σ 0 := ⟨fun _ => 0, by norm_num⟩
+
+def Exponent.one (i : σ) : Exponent σ 1 := ⟨fun j => if i = j then 1 else 0, by norm_num⟩
+
+instance exponentFinite (n : ℕ) : Finite (Exponent σ n) := by
+  let f : Exponent σ n → (σ → Fin (n+1)) := fun s => fun i => ⟨s.exp i, Nat.lt_succ_of_le (Exponent.single_le_sum σ s i)⟩
+  have hf : Function.Injective f := by
+    intros s t h
+    rw [Exponent.ext_iff]
+    ext i
+    have := congr_arg (fun s => s i) h
+    simp only [Function.Injective.eq_iff] at this
+    exact Fin.mk_eq_mk.mp this
+  exact Finite.of_injective f hf
+  done
+
+instance exponentFintype (n : ℕ) : Fintype (Exponent σ n) := Fintype.ofFinite (Exponent σ n)
+
+def exponentLiftEquiv (e : σ ≃ τ) : Exponent σ n ≃ Exponent τ n := {
+          toFun := fun s => ⟨fun i => s.exp (e.invFun i), by simp_rw [← s.sum_eq]; exact Equiv.sum_comp e.symm s.exp⟩,
+          invFun := fun s => ⟨fun i => s.exp (e.toFun i), by simp_rw [← s.sum_eq]; exact Equiv.sum_comp e s.exp⟩,
+          left_inv := fun s => by simp,
+          right_inv := fun s => by simp
+          }
+
+lemma exponentLiftEquiv_toFun (e : σ ≃ τ) (s : Exponent σ n) (i : τ) : ((exponentLiftEquiv _ e).toFun s).exp i = s.exp (e.invFun i) := by
+  simp only [Equiv.toFun_as_coe, Equiv.invFun_as_coe]
+  rfl
+  done
+
+lemma Exponent.exponentCard_zero : Fintype.card (Exponent σ 0) = 1 := by
+  apply Fintype.card_eq_one_iff.mpr
+  use Exponent.zero σ
+  intro x
+  ext i
+  simp only [Exponent.sum_eq, Exponent.zero_of_sum_eq_zero]
+  done
+
+lemma Exponent.one_sum_singleton (s : Exponent σ 1) : ∃! (i : σ) , Exponent.one σ i = s:= by
+  have h : ∑ i, s.exp i = 1 := s.sum_eq
+  have (i : σ): exp s i = 0 ∨ exp s i = 1 := by
```

Comment:
As a meta comment: this proof is (relatively) long and deeply nested. Usually, that means it can be simplified or split up.
In other words, it could help to ask yourself questions like
- are there intermediate results proven that will be useful outside of this lemma?
- can you simplify ("golf") the proof?
- do you need all the intermediate `have`s: can they be inlined or dropped?
(I'd have to know your field better to answer any of these conclusively. Addressing the style comments will certainly help with the second point; often, this also makes the other two points clearer.)